File: flink-examples/flink-examples-table/src/main/java/org/apache/flink/table/examples/java/basics/StreamIntervalJoinSQLExample.java
package org.apache.flink.table.examples.java.basics;

/**
 * @author haxi
 * @description
 * @date 2024/4/5 16:47
 */

import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.ExplainDetail;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class StreamIntervalJoinSQLExample {

    private static final Logger logger = LoggerFactory.getLogger(StreamIntervalJoinSQLExample.class);
    public static void main(String[] args) throws Exception {

        // set up the Java DataStream API
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // set up the Java Table API
        final StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

        String left = "CREATE TABLE left_source_table (\n"
                + "    order_id STRING,\n"
                + "    user_id STRING,\n"
                + "    price BIGINT,\n"
                + "    proctime AS PROCTIME()\n"
                + ") WITH (\n"
                + "  'connector' = 'datagen',\n"
                + "  'rows-per-second' = '10',\n"
                + "  'fields.order_id.length' = '1',\n"
                + "  'fields.user_id.length' = '10',\n"
                + "  'fields.price.min' = '1',\n"
                + "  'fields.price.max' = '1000000'\n"
                + ")";

        String right = "CREATE TABLE right_source_table (\n"
                + "    order_id STRING,\n"
                + "    user_id STRING,\n"
                + "    price BIGINT,\n"
                + "    proctime AS PROCTIME()\n"
                + ") WITH (\n"
                + "  'connector' = 'datagen',\n"
                + "  'rows-per-second' = '10',\n"
                + "  'fields.order_id.length' = '1',\n"
                + "  'fields.user_id.length' = '10',\n"
                + "  'fields.price.min' = '1',\n"
                + "  'fields.price.max' = '1000000'\n"
                + ")";

        String query = "select l.order_id, l.user_id, r.order_id, r.user_id, l.price, r.price "
                + "from left_source_table l, right_source_table r "
                + "where l.order_id = r.order_id "
                + "and l.proctime BETWEEN r.proctime - INTERVAL '10' SECOND AND r.proctime + INTERVAL '5' SECOND";


        tableEnv.executeSql(left);
        tableEnv.executeSql(right);
        logger.info(tableEnv.explainSql(query, ExplainDetail.CHANGELOG_MODE));
    }
}

File: flink-examples/flink-examples-table/src/main/scala/org/apache/flink/table/examples/scala/basics/StreamIntervalJoinSQLExample.scala
package org.apache.flink.table.examples.scala.basics

import org.apache.flink.api.scala._
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.table.api.{FieldExpression, Schema, TableEnvironment, UnresolvedFieldExpression}
import org.apache.flink.table.api.bridge.scala.{dataStreamConversions, tableConversions, StreamTableEnvironment}
import org.apache.flink.types.Row

import java.sql.Timestamp
import java.time.ZoneId

import scala.collection.mutable

object StreamIntervalJoinSQLExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    val tEnv = StreamTableEnvironment.create(env)
    tEnv.getConfig.setLocalTimeZone(ZoneId.of("Asia/Shanghai"))
//    env.setParallelism(1)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    // order data
    val ordersData = new mutable.MutableList[(String, String, Timestamp)]
    ordersData.+=(("001", "iphone", new Timestamp(1545800002000L)))
    ordersData.+=(("002", "mac", new Timestamp(1545800003000L)))
    ordersData.+=(("003", "book", new Timestamp(1545800004000L)))
    ordersData.+=(("004", "cup", new Timestamp(1545800018000L)))

    // payment data
    val paymentData = new mutable.MutableList[(String, String, Timestamp)]
    paymentData.+=(("001", "alipay", new Timestamp(1545803501000L)))
    paymentData.+=(("002", "card", new Timestamp(1545803602000L)))
    paymentData.+=(("003", "card", new Timestamp(1545803610000L)))
    paymentData.+=(("004", "alipay", new Timestamp(1545803611000L)))

    // schema
    val orders = env
      .fromCollection(ordersData)
      .assignTimestampsAndWatermarks(new TimestampExtractor())
      .toTable(tEnv, $"orderId", $"productName", $"orderTime")

    val ratesHistory = env
      .fromCollection(paymentData)
      .assignTimestampsAndWatermarks(new TimestampExtractor())
      .toTable(tEnv, $"orderId", $"payType", $"payTime")

    tEnv.createTemporaryView("Orders", orders)
    tEnv.createTemporaryView("Payment", ratesHistory)

    var sqlQuery =
      """
        |SELECT
        |  o.orderId,
        |  o.productName,
        |  p.payType,
        |  o.orderTime,
        |  cast(payTime as timestamp) as payTime
        |FROM
        |  Orders AS o JOIN Payment AS p ON o.orderId = p.orderId AND
        | p.payTime BETWEEN orderTime AND orderTime + INTERVAL '1' HOUR
        |""".stripMargin
//    tEnv.explainSql(sqlQuery)
    tEnv.executeSql(sqlQuery).print()
//    tEnv.createTemporaryView("TemporalJoinResult", tEnv.sqlQuery(sqlQuery))

//    val result = tEnv.from("TemporalJoinResult").toChangelogStream
//    result.print()
    env.execute()
  }
}

class TimestampExtractor
  extends BoundedOutOfOrdernessTimestampExtractor[(String, String, Timestamp)](Time.seconds(10)) {
  override def extractTimestamp(element: (String, String, Timestamp)): Long = {
    element._3.getTime
  }
}
